hubblestack/extmods/fdg/cert_discovery.py

```diff
@@ -0,0 +1,188 @@
+        cert_details['serial_number'] = str(x509.get_serial_number())
+        cert_details['expiry_date'] = str(not_after)
+        cert_details['issue_date'] = str(not_before)
+        cert_details['signature_algo'] = str(x509.get_signature_algorithm())
+        cert_details['pem_cert'] = str(cert['data'])
+        cert_details['SAN'] = get_certificate_san(x509) 
+    except Exception as e:
+        cert_details['error'] = "some error occurred while parsing certificate - {0}".format(e)
+    return cert_details
+
+def fill_na(host, port, message):
+    """
+    Fill 'NA' in case of 'no cert found' on the input port.
+    """
+    cert_details = {}
+    cert_details['country_name'] = 'NA'
+    cert_details['organisation_name'] = 'NA'
+    cert_details['organisation_unit_name'] = 'NA'
+    cert_details['common_name'] = 'NA'
+    cert_details['version'] = 'NA'
+    cert_details['has_expired'] = 'NA'
+    cert_details['serial_number'] = 'NA'
+    cert_details['issuer'] = 'NA'
+    cert_details['expiry_date'] = 'NA'
+    cert_details['issue_date'] = 'NA'
+    cert_details['signature_algo'] = 'NA'
+    cert_details['pem_cert'] = 'NA'
+    cert_details['dest_port'] = str(port)
+    cert_details['error'] = message
+    cert_details['dest_ip'] = str(host)
+
+    return cert_details
+
+def get_cert_details(host_ip='', host_port='', chained=None, chained_status=None):
+    """
+    This module is used to fetch certificate details on a host and port.
+    This module can also be used in conjunction with osquery as the first module
+    in the chain. Given that osquery fetches information about the open
+    ports on a system and provides a 'host, port' tuple (or a list of host, port tuples)
+    to this module, this module will connect to the host and port and fetch
+    certificate details if a certificate is attached on the port. As an example,
+    Osquery needs to provide the value of 'chained' in the following format.
+    +-------------------------------+-----------+
+    | host_ip                       | host_port |
+    +-------------------------------+-----------+
+    | 127.0.0.1                     | 80        |
+    | 2001:db8:85a3::8a2e:370:7334  | 80        |
+    | 127.0.0.1                     | 443       |
+    | 2001:db8:85a3::8a2e:370:7334  | 443       |
+    +-------------------------------+-----------+
+    The first return value (status) will be True if the module is able to
+        1. Connect to the port and fetch certificate details.
+        2. Connect to the port and exit if no certificate is attached on the port.
+    The first return value (status) will be False if the module encounters some
+    exception in python's ssl.get_server_certificate function.
+
+    chained
+        The value chained from the previous call.
+
+    chained_status
+        The status returned by the chained call.
+    """
+    if host_ip == "":
+        host = str(chained.get('host_ip', ''))
+    else:
+        host = str(host_ip.get('host_ip', ''))
```

Comment:
You still have if else block :)  This can be done in one line expression:  `host = str(chained.get('host_ip', host_ip.get('host_ip', '')))`

Same for port